compiler/rustc_hir_typeck/src/fn_ctxt/checks.rs

```diff
@@ -1447,6 +1451,27 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             }
             self.diverges.set(previous_diverges);
         }
+
+        self.lint_unit_bindings(decl_ty, init_ty, decl);
+    }
+
+    fn lint_unit_bindings(
+        &self,
+        decl_ty: Ty<'tcx>,
+        init_ty: Option<Ty<'tcx>>,
+        decl: Declaration<'tcx>,
+    ) {
+        if decl.span.can_be_used_for_suggestions()
+            && let Some(init_ty) = init_ty
+            && self.infcx.can_eq(self.param_env, init_ty, self.tcx.mk_unit())
+            && decl.ty.is_none() // suppress if user explicitly ascribes a type to the pattern
+            && self.infcx.can_eq(self.param_env, decl_ty, self.tcx.mk_unit())
+            && !matches!(decl.pat.kind, hir::PatKind::Wild) // suppress if user writes `let _ = expr;`.
```

Comment:
why is this necessary? `let _ = unit()` seems redundant too.

Reply:
> why is this necessary?

Sorry which one? Do you mean "suppress if user explicitly ascribes a type to the pattern"?

> `let _ = unit()` seems redundant too.

I will remove this case.